packages/shorebird_cli/lib/src/artifact_builder/flutter_build_process_tracker.dart

```diff
@@ -0,0 +1,136 @@
+  downloadingGradleDependency,
+
+  /// Flutter is checking and preparing the Android SDK. (Android only)
+  preparingAndroidSDK,
+
+  /// The flutter assemble command is executing.
+  flutterAssemble;
+
+  /// A message that describes the step.
+  ///
+  /// When a step message should be defined by the user interface, this
+  /// method will return null.
+  String? get message {
+    switch (this) {
+      case FlutterBuildProcessTrackerStep.downloadingGradleW:
+        return 'Downloading Gradle Wrapper';
+      case FlutterBuildProcessTrackerStep.downloadingGradleDependency:
+        return 'Downloading Gradle Dependency';
+      case FlutterBuildProcessTrackerStep.preparingAndroidSDK:
+        return 'Preparing Android SDK';
+      case FlutterBuildProcessTrackerStep.flutterAssemble:
+        return 'Assembling Flutter Resources';
+      case FlutterBuildProcessTrackerStep.initial:
+      case FlutterBuildProcessTrackerStep.building:
+        return null;
+    }
+  }
+}
+
+/// {@template flutter_build_log_updater}
+///
+/// FlutterBuildProcessTracker is a class that will process the build logs
+/// from a verbose flutter build command and break that down into different
+/// steps in order to give a better sense of progress to the user.
+///
+/// {@endtemplate}
+class FlutterBuildProcessTracker extends ShorebirdProcessTracker {
+  /// {@macro flutter_build_log_updater}
+  FlutterBuildProcessTracker({
+    required this.baseMessage,
+    Progress? progress,
+  }) : progress = progress ?? logger.progress(baseMessage);
+
+  /// The overlaying message of the build process.
+  final String baseMessage;
+
+  /// The progress object that will update the user interface.
+  late final Progress progress;
+
+  final _steps = [FlutterBuildProcessTrackerStep.initial];
+
+  /// The steps that have been identified in the build process.
+  List<FlutterBuildProcessTrackerStep> get steps =>
+      UnmodifiableListView(_steps);
+
+  /// The current step in the build process.
+  FlutterBuildProcessTrackerStep get currentStep => _steps.last;
+
+  @override
+  void onLog(String logLine) {
+    super.onLog(logLine);
+    if (logLine.contains("Running Gradle task 'bundleRelease'...") &&
+        currentStep == FlutterBuildProcessTrackerStep.initial) {
+      _emitStep(FlutterBuildProcessTrackerStep.building);
+    } else if (logLine.contains(
+      'Downloading https://services.gradle.org/distributions/gradle',
+    )) {
+      _emitStep(FlutterBuildProcessTrackerStep.downloadingGradleW);
+    } else if (logLine.contains('Downloading') && logLine.contains('.gradle')) {
+      _emitStep(FlutterBuildProcessTrackerStep.downloadingGradleDependency);
+    } else if (logLine.contains('Parsing') &&
+        logLine.contains('android${platform.pathSeparator}sdk') &&
+        currentStep != FlutterBuildProcessTrackerStep.preparingAndroidSDK) {
+      _emitStep(FlutterBuildProcessTrackerStep.preparingAndroidSDK);
+    } else if (logLine.contains('SDK initialized in') &&
```

Comment:
It seems like some of these care about step order and some of them don't?

Reply:
Yes. It is more about "starting and finishing an step" instead of the order itself.

For example the following:

```
 } else if (logLine.contains('Parsing') &&
        logLine.contains('android${platform.pathSeparator}sdk') &&
        currentStep != FlutterBuildProcessTrackerStep.preparingAndroidSDK) {
      _emitStep(FlutterBuildProcessTrackerStep.preparingAndroidSDK);
```

There isn't any obvious log like "Starting to install Android SDK" that we can look for, but I realized from testing that the preparing android SDK step starts when we start to see "Parsing ... android/sdk". But the log will have many lines like that in the sequence, so this logic will make sure that we trigger the `preparingAndroidSDK` step only when we see the first line of such message.